sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+# Shared DPC++ libraries
+
+This document describes purpose and design of Shared DPC++ libraries feature.
+
+## Background
+Sometimes users want to provide *device* functions via shared libraries.
+Simple source example:
+```
+// App:
+
+CGH.parallel_for<app_kernel>(/* ... */ {
+  library_function();
+});
+
+
+// Shared library:
+SYCL_EXTERNAL void library_function() {
+  // do something
+}
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+The main purpose of this feature is to provide a mechanism which allows to
+provide *device* functions via shared libraries and works as close as possible
+to regular shared libraries.
+
+## Requrements:
+User's code is compiled into a shared library which consists of some host API,
+device code and device API (`SYCL_EXTERNAL` functions). The library is linked to
+a user's application which also contains some device code and performs
+computations using DPC++/SYCL.
+For this combination the following statements must be true:
+
+- `SYCL_EXTERNAL` functions from library can be called (directly or indirectly)
+  from device code of the application.
+- Function pointers taken in application should work inside the library.
+- Specific code changes are not required, i.e. the mechanism of linking works
+  as close as possible to regular shared libraries.
+
+## Design
+The overall idea is simple:
+
+- Each device image is supplied with an information about exported and imported
+  symbols using device image properties
+- DPC++ RT performs *device images collection* task by grouping all device
+  images required to execute a kernel based on the list of exports/imports
+  - Besides symbol names, additional attributes are taken into account (like
+    device image format: SPIR-V or device asm)
+- Actual linking is performed by underlying backend (OpenCL/L0/etc.)
+
+Next sections describe details of changes in each component.
+
+### DPC++ front-end changes
+
+DPC++ front-end generates `module-id` attribute on each `SYCL_EXTERNAL` function.
+It was generated only on kernels earlier. There are two reasons to start
+generating this attribute on `SYCL_EXTERNAL` functions:
+
+- Later in pipeline, this attribute will be used by `sycl-post-link` tool to
+  separate `SYCL_EXTERNAL` functions from non-`SYCL_EXTERNAL` functions with
+  external linkage.
+- `module-id` attribute also contains information about source file where the
+  function comes from. This information will be used to perform device code
+  split on device images that contain only exported functions.
```

Comment:
it is not clear what is module-id - could you please specify? Is this a source file only? I guess that would not be enough. Same device function can appear in multiple device images with the same source file attribute,

Reply:
> I guess that would not be enough. Same device function can appear in multiple device images with the same source file attribute,

If that is a problem, I suggest to solve it separately, because `module-id` mechanism is just being re-used from device code split implementation

Reply:
I don't know if this can be a real problem. We know that using library file name (.dll/.so) as the module ID works, but it is not clear whether using just source name can be as safe. I guess, I'd wait for the description of export/import lists and RT resolution I requested above, maybe this will address concerns about module-id.

Reply:
`module-id` is just a name of source file where the function was defined. It is used only for device code split mechanism. RT doesn't use this information for linking. I expanded RT changes description.